lib/dao.js

```diff
@@ -630,7 +630,173 @@ DataAccessObject.upsert = function(data, options, cb) {
+  assert(typeof cb === 'function', 'The cb argument must be a function');
+  if (Object.keys(data).length === 0) {
+    var err = new Error('data object cannot be empty!');
+    err.statusCode = 400;
+    process.nextTick(function() { cb(err); });
+    return cb.promise;
+  }
+  var hookState = {};
+  var self = this;
+  var Model = this;
+  var connector = Model.getConnector();
+  var modelName = Model.modelName;
+  var query = { where: where };
+  var context = {
+    Model: Model,
+    query: query,
+    hookState: hookState,
+    options: options,
+  };
+  Model.notifyObserversOf('access', context, doUpsertWithWhere);
+  function doUpsertWithWhere(err, ctx) {
+    if (err) return cb(err);
+    ctx.data = data;
+    if (connector.upsertWithWhere) {
+      var context = {
+        Model: Model,
+        where: ctx.query.where,
+        data: ctx.data,
+        hookState: hookState,
+        options: options,
+      };
+      Model.notifyObserversOf('before save', context, function(err, ctx) {
+        if (err) return cb(err);
+        data = ctx.data;
+        var update = data;
+        var inst = data;
+        if (!(data instanceof Model)) {
+          inst = new Model(data, { applyDefaultValues: false });
+        }
+        update = inst.toObject(false);
+        Model.applyScope(query);
+        Model.applyProperties(update, inst);
+        Model = Model.lookupModel(update);
+        if (options.validate === false) {
+          return callConnector();
+        }
+        if (options.validate === undefined && Model.settings.automaticValidation === false) {
+          return callConnector();
+        }
+        inst.isValid(function(valid) {
+          if (!valid) return cb(new ValidationError(inst), inst);
+          callConnector();
+        }, update, options);
 
+        function callConnector() {
+          try {
+            ctx.where = removeUndefined(ctx.where);
+            ctx.where = Model._coerce(ctx.where);
+            update = removeUndefined(update);
+            update = Model._coerce(update);
+          } catch (err) {
+            return process.nextTick(function() {
+              cb(err);
+            });
+          }
+          context = {
+            Model: Model,
+            where: ctx.where,
+            data: update,
+            currentInstance: inst,
+            hookState: ctx.hookState,
+            options: options,
+          };
+          Model.notifyObserversOf('persist', context, function(err) {
+            if (err) return done(err);
+            connector.upsertWithWhere(modelName, ctx.where, update, options, done);
+          });
+        }
+        function done(err, data, info) {
+          if (err) return cb(err);
+          var contxt = {
+            Model: Model,
+            data: data,
+            isNewInstance: info && info.isNewInstance,
+            hookState: ctx.hookState,
+            options: options,
+          };
+          Model.notifyObserversOf('loaded', contxt, function(err) {
```

Comment:
ummm not sure why you could not use `context` as opposed to `contxt`; probably it conflicts the name. How about using `ctx` instead of `contxt` which is missing **e**